### PR TITLE
Variational autoencoder for image-based profiles. 

### PR DESCRIPTION
**Pull request by axiomcura**
In this PR, you'll find the Jupyter Notebook showcasing the development of the Variational Autoencoder (VAE) for image-based profiles, along with additional extracted files tailored for 'pre-commit' integration and environment file.

### CODE REVIEW

File: notebooks/nbconverted/VAE_test.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+# training oop
+for epoch in range(num_epochs):
+    for data in dataloader:
+        optimizer.zero_grad()
+        x = data
+        x_recon, mean, logvar = vae(x)
+        loss = vae_loss(x, x_recon, mean, logvar)
+        loss.backward()
+        optimizer.step()
+    print(f"Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}")
```
Comment (MattsonCam) [inline notebooks/nbconverted/VAE_test.py:332]: This looks like it is tracking the training loss, also consider incorporating the validation loss

File: notebooks/nbconverted/VAE_test.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+# Initialize VAE and optimizer
+vae = ImageProfileVAE(input_dim, latent_dim)
+optimizer = optim.Adam(vae.parameters(), lr=0.001)
+
+
+# In[21]:
+
+
+# training oop
+for epoch in range(num_epochs):
```
Comment (MattsonCam) [inline notebooks/nbconverted/VAE_test.py:324]: In the future, also consider adding hyperparameter optimization as well and other ways of improving the model learning

File: notebooks/nbconverted/VAE_test.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+# The `CustomDataset` class is designed to facilitate the integration of Pandas DataFrames with PyTorch for deep learning tasks.
+# It converts the data from a Pandas DataFrame into a PyTorch Tensor with the appropriate data type, allowing for efficient usage in PyTorch data loaders.
+# The class includes methods for determining the dataset's length and retrieving data samples by index, making it suitable for structured data applications.
+# A typical use case is to create a `CustomDataset` instance from a Pandas DataFrame, enabling data loading for training and inference in PyTorch deep learning models.
+#
+
+# In[3]:
+
+
+class ImageProfileDataset(Dataset):
```
Comment (MattsonCam) [inline notebooks/nbconverted/VAE_test.py:192]: I've used Dataset briefly, there may be a better way to do this to remove one-line functions

File: notebooks/nbconverted/VAE_test.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+        )
+
+        # define the layers responsible for mapping the output of the encoder neural network
+        self.fc_mean = nn.Linear(32, latent_dim)
+        self.fc_logvar = nn.Linear(32, latent_dim)
+
+        # building the decoder sequence
+        self.decoder = nn.Sequential(
+            nn.Linear(latent_dim, 32),
+            nn.ReLU(),
```
Comment (MattsonCam) [inline notebooks/nbconverted/VAE_test.py:56]: ReLU is a good default activation function to use, this may be worth tuning later

File: notebooks/nbconverted/VAE_test.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+            nn.Linear(32, 64),
+            nn.ReLU(),
+            nn.Linear(64, input_dim),
+            nn.Sigmoid(),
```
Comment (MattsonCam) [inline notebooks/nbconverted/VAE_test.py:60]: I would probably change this to a  linear activation function as the output instead. What is the range of your features?

**Review by MattsonCam — CHANGES_REQUESTED** [review]
Great job Erik! I had one comment about the output activation function I think should be addressed, otherwise, nice work!

File: notebooks/nbconverted/VAE_test.py (hunk 4 above)
Comment (axiomcura) [reply notebooks/nbconverted/VAE_test.py:56]: Yes! However, this will change in the future after some reseaching and testing!

File: notebooks/nbconverted/VAE_test.py (hunk 5 above)
Comment (axiomcura) [reply notebooks/nbconverted/VAE_test.py:60]: What is the range? I am assuming you are asking how many features are within the dataset. probably between 1000~2000 features, since the features were extracted by using CellProfiler

File: notebooks/nbconverted/VAE_test.py (hunk 3 above)
Comment (axiomcura) [reply notebooks/nbconverted/VAE_test.py:192]: That's a good point! I used the `Dataset` class in order for the plate data can be formated in a way for Pytorch to take in as an input.

File: notebooks/nbconverted/VAE_test.py (hunk 2 above)
Comment (axiomcura) [reply notebooks/nbconverted/VAE_test.py:324]: Absolutely, @MikeLippincott and I are planning to do an Optuna hyperparameter search as soon as we fine-tune our model for optimal efficiency!

File: notebooks/nbconverted/VAE_test.py (hunk 1 above)
Comment (axiomcura) [reply notebooks/nbconverted/VAE_test.py:332]: most def! I will create an issue

**Comment by axiomcura** [comment]
@MattsonCam Thanks for your comments and suggestions. Created some new issues. Merging now!
